base/docs/Docs.jl

```diff
@@ -285,41 +272,36 @@ function objdoc(meta, def)
     end
 end
 
-fexpr(ex) = isexpr(ex, :function, :(=)) && isexpr(ex.args[1], :call)
-
 function docm(meta, def)
-    # Quote, Unblock and Macroexpand
-    # * Always do macro expansion unless it's a quote (for consistency)
-    # * Unblock before checking for Expr(:quote) to support `->` syntax
-    # * Unblock after macro expansion to recognize structures of
-    #   the generated AST
-    def′ = unblock(def)
-    if !isexpr(def′, :quote)
-        def = macroexpand(def)
-        def′ = unblock(def)
-    elseif length(def′.args) == 1 && isexpr(def′.args[1], :macrocall)
-        # Special case for documenting macros after definition with
-        # `@doc "<doc string>" :@macro` or
-        # `@doc "<doc string>" :(str_macro"")` syntax.
-        #
-        # Allow more general macrocall for now unless it causes confusion.
-        return objdoc(meta, namify(def′.args[1]))
-    end
-    isexpr(def′, :macro) && return namedoc(meta, def, symbol("@", namify(def′)))
-    isexpr(def′, :type) && return typedoc(meta, def, namify(def′.args[2]))
-    isexpr(def′, :bitstype) && return namedoc(meta, def, def′.args[2])
-    isexpr(def′, :abstract) && return namedoc(meta, def, namify(def′))
-    isexpr(def′, :module) && return namedoc(meta, def, def′.args[2])
-    fexpr(def′) && return funcdoc(meta, def)
-    return objdoc(meta, def)
+    @match def begin
+        :(@m_)  -> return objdoc(meta, m)
+          m_""  -> return objdoc(meta, m)
+    end
+    def = macroexpand(def)
+    @match def begin
+        (f_(__) = _)          -> funcdoc(meta, def)
+        function f_(__) _ end -> funcdoc(meta, def)
+        function f_ end       ->  objdoc(meta, def)
+        macro m_(__) _ end    -> namedoc(meta, def, symbol("@", m))
+        type T_ _ end         -> typedoc(meta, def, namify(T))
+        immutable T_ _ end    -> typedoc(meta, def, namify(T))
+        (abstract T_)         -> namedoc(meta, def, namify(T))
+        (bitstype _ T_)       -> namedoc(meta, def, namify(T))
+        (typealias T_ _)      ->  objdoc(meta, def)
+        module M_ _ end       -> namedoc(meta, def, M)
+        _                     -> !isa(def, Expr) ? objdoc(meta, def) :
+                                     error("Unsupported @doc syntax $def")
```

Comment:
I disabled this for now because its behaviour (documenting the value `1`) is confusing / prone to error. I'll re-enable once I've implemented docs for bindings.